libraries/common/stream.effekt

```diff
@@ -0,0 +1,234 @@
+module stream
+
+import bytes
+
+interface Emit[A] {
+  def emit(value: A): Unit
+}
+
+def for[A, R] { stream: () => R / Emit[A] } { action: A => Unit }: R =
+  try {
+    stream()
+  } with Emit[A] {
+    def emit(value) = resume(action(value))
+  }
+
+// TODO what if the number is negative? undefined behavior?
+def repeat[A](number: Int) { action: () => A }: Unit / Emit[A] =
+  if (number > 0) {
+    do emit(action())
+    repeat(number - 1) {action}
+  }
+
+
+def each[A](list: List[A]): Unit / Emit[A] =
+  list match {
+    case Nil() => ()
+    case Cons(head, tail) =>
+      do emit(head)
+      each(tail)
+  }
+
+def each[A](array: Array[A]): Unit / Emit[A] = {
+  val n = array.size
+  def go(i: Int): Unit = {
+    if (i < n) {
+      do emit(array.unsafeGet(i))
+      go(i + 1)
+    }
+  }
+  go(0)
+}
+
+def each(bytes: Bytes): Unit / Emit[Byte] = {
+  val n = bytes.size
+  def go(i: Int): Unit = {
+    if (i < n) {
+      do emit(bytes.read(i))
+      go(i + 1)
+    }
+  }
+  go(0)
+}
+
+// not option
+// not dequeue
+// not queue
+// not result
+// not seq
+
+// def decodeUtf8[R] { stream: () => R / Emit[Byte] }: R / Emit[Char] = <>
+// def encodeUtf8[R] { stream: () => R / Emit[Char] }: R / Emit[Byte] = <>
+// def each(string: String): Unit / Emit[Char] = decodeUtf8 { string.asBytes.each }
+
+
+interface Done {
+  def done(): Nothing
+}
+
+def noneOnDone[R] { program: () => R / Done }: Option[R] =
+  try {
+    Some(program())
+  } with Done {
+    def done() = None()
+  }
+
+
+record Indexed[A](index: Int, value: A)
+
+// In Effekt lower bounds are inclusive and upper bounds are exclusive
```

Comment:
Same for this, this should be in `effekt.effekt` somewhere in the toplevel documentation.